### ISSUE TITLE
Use player controls on iOS lock screen

### ISSUE THREAD

**Issue by dimdieo** [libpd/libpd#375]
I'm trying to get music playback control on the lock screen, but I'm not getting anything. If this function is supported in libPD? I'm trying to interact with MPNowPlayingInfoCenter, but libPD doesn't seem to support it directly

**Comment by danomatika** [comment]
If you are using iOS, you need to enable the background audio entitlement. LibPd doesn't do this for you automatically.

**Comment by dimdieo** [comment]
Thanks for the answer. Yes, I did it first. I am using SwiftUI, for PdEngine.swift I did this

```
class MyPdAudioController: PdAudioController {
    
    override func configurePlayback(withSampleRate sampleRate: Int32, numberChannels numChannels: Int32, inputEnabled: Bool, mixingEnabled: Bool) -> PdAudioStatus {
        var status = super.configurePlayback(withSampleRate: sampleRate, numberChannels: numChannels, inputEnabled: inputEnabled, mixingEnabled: mixingEnabled)
        
        if !MyPdAudioController.setSessionOptions(currentCategoryOptions()) {
            status = PdAudioError
            
        }
        return status
    }
    
    override class func setSessionOptions(_ options: AVAudioSession.CategoryOptions) -> Bool {
        let session = AVAudioSession.sharedInstance()

        do {
            try session.setCategory(.playAndRecord, mode: .spokenAudio, options: [.defaultToSpeaker, .allowAirPlay, .allowBluetoothA2DP])
            try session.setActive(true)
            
            return true
        } catch let error {
            Swift.print("Failed to set audio session category. Error: \(error)")
            return false
            
        }
    }

}

class PdEngine {
    @Published var isPlaying: Bool = false
    static let shared = PdEngine()
    var pd: MyPdAudioController?
    var nowPlayingInfo = [String : Any]()
    
    private init() {
        pd = MyPdAudioController()
        let sampleRate = getSampleRateForDevice()
        configureBufferSize(ticksPerBuffer: 16)
        pd?.configurePlayback(withSampleRate: Int32(sampleRate), inputChannels: 2, outputChannels: 2, inputEnabled: true)
        ffplay_tilde_setup()
        rand_setup()
        pd?.isActive = true
        
        setupRemoteTransportControls()
        
    }
    
    // Update the Now Playing Info Center
    func setupNowPlayingInfo() {
        nowPlayingInfo[MPMediaItemPropertyTitle] = "Test Title"
        nowPlayingInfo[MPNowPlayingInfoPropertyElapsedPlaybackTime] = 0
        nowPlayingInfo[MPMediaItemPropertyPlaybackDuration] = 100
        MPNowPlayingInfoCenter.default().nowPlayingInfo = nowPlayingInfo
    }

    func setupRemoteTransportControls() {
        let commandCenter = MPRemoteCommandCenter.shared()
        
        commandCenter.playCommand.addTarget { [unowned self] event in
            if self.isPlaying == false {
                self.play()
                return .success
            }
            return .commandFailed
        }
        commandCenter.playCommand.isEnabled = true

        commandCenter.pauseCommand.addTarget { [unowned self] event in
            if self.isPlaying == true {
                self.pause()
                return .success
            }
            return .commandFailed
        }
        commandCenter.pauseCommand.isEnabled = true
    }


    func play() {
        pd?.isActive = true
        isPlaying = true
        setupNowPlayingInfo()
    }

    func pause() {
        pd?.isActive = false
        isPlaying = false
    }
    
// other code
```
    
But still no effect.

**Comment by danomatika** [comment]
You shouldn't need to subclass PdAudioController. In any case, this isn't really a LibPd issue itself, but a general AVAudioSession on iOS issue so it's not something I can directly answer quickly here. Are the events received, ie. add prints? Also, the pd audio controller is active but have you enabled DSP? -> `PdBase.computerAudio(true)`. If not, you might be receiving the events but no audio is being computed...

**Comment by dimdieo** [comment]
I added a subclass to try changing the AVAudioSession. The fact is that if you use AVAudioPlayer to simply play a file and specify

`try AVAudioSession.sharedInstance().setCategory(.playback, mode: .default, options: [.duckOthers, .interruptSpokenAudioAndMixWithOthers])`

I don't get control buttons either, but if I specify 

`try AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .spokenAudio, options: [.defaultToSpeaker, .allowAirPlay, .allowBluetoothA2DP])`

then the buttons are in place.

**Comment by danomatika** [comment]
Ok, that makes sense. I have only set up media control buttons in apps *without* LibPd, so I can't personally say much more. It's the usual problem of finding the *specific* settings the system requires for it to work...

**Comment by dimdieo** [comment]
I solved the problem in the following way.
 
```
 private init() {
        pd = PdAudioController()
        private init() {
        pd = PdAudioController()
        let sampleRate = getSampleRateForDevice()
        configureBufferSize(ticksPerBuffer: 16)
        pd?.configurePlayback(withSampleRate: Int32(sampleRate), inputChannels: 2, outputChannels: 2, inputEnabled: true)
        
        setupAVAudioSession() // add new function
        
        pd?.isActive = true
        
    }
    
    func setupAVAudioSession() {
        do {
            try AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .spokenAudio, options: [.defaultToSpeaker, .allowAirPlay, .allowBluetoothA2DP])
            try AVAudioSession.sharedInstance().setActive(true)
        } catch {
            print("Failed to set up AVAudioSession: \(error)")
        }
    }
```
    
    After that, the playback control item finally appeared on the lock screen. Maybe it will be useful for someone

**Comment by danomatika** [comment]
I wonder, does setting the options and mode through the controller work? This should basically do the same thing. Something like:

~~~
pd = PdAudioController()
pd?.defaultToSpeaker = true
pd?.allowAirPlay = true
pd?.allowBluetoothA2DP = true
pd?.mode = .spokenWord
let sampleRate = getSampleRateForDevice()
configureBufferSize(ticksPerBuffer: 16)
pd?.configurePlayback(withSampleRate: Int32(sampleRate), inputChannels: 2, outputChannels: 2, inputEnabled: true)
~~~

**Comment by dimdieo** [comment]
Unfortunately it is not possible to activate the control center on the lock screen in this way.

These values are set without error

```
pd?.defaultToSpeaker = true
pd?.allowAirPlay = true
pd?.allowBluetoothA2DP = true
```

But for the "pd?.mode" I can't put any value be it .spokenAudio or .default
	
I tried to run without "pd?.mode", but it doesn't activate. 


I also noticed that if I set the "setupAVAudioSession()" after "pd = PdAudioController()" the control center is also not activated.

At the same time, I cannot solve the problem: if I stop at the solution that I suggested above, then the application starts with the Play button in the control center on the locked screen. At the same time, when you click on the Play button in the application, the Play button in the control center does not react in any way. Although if you press the Play button in the control center, then the button presses are synchronized. I can't figure out what could be the issue. It's a shame that the application is already ready after several months of work and I have come across this problem that I can not solve in any way.

**Comment by danomatika** [comment]
> But for the "pd?.mode" I can't put any value be it .spokenAudio or .default

It looks like you have to convert from the string enum value to a string:
~~~
pd?.mode = AVAudioSession.Mode.spokenAudio as NSString
~~~

This comes from the classes being written in Obj-C.

As for the rest, I can't say. It sounds like a synchronization issue between the interface elements and not something specific to libpd.

**Comment by dimdieo** [comment]
Thanks for the reply, the error is gone. But it looks like you need to set a category that has these options. That's how it is here

`try AVAudioSession.sharedInstance().setCategory(.playAndRecord, mode: .default, options: [.defaultToSpeaker, .allowAirPlay, .allowBluetoothA2DP])`

What is the best way to do this?

**Comment by danomatika** [comment]
`pd?.configurePlayback()` with input & output channels as well as inputEnabled = true should set the .playAndRecord category for you. If you have the mode and options set beforehand, they are then applied. Basically, the PdAudioController wraps setting up the AVAudioSession: https://github.com/libpd/libpd/blob/master/objc/PdAudioController.m#L531

**Comment by dimdieo** [comment]
Unfortunately, this approach doesn't work properly for some reason, but it works through the setupAVAudioSession() function call.
And this function must be after "pd?.configurePlayback" otherwise, if it is placed after the call to "pd=PdAudioController()" then the control center on the lock screen is not activated. I also tried moving the lines "pd?." in the same place where the function setupAVAudioSession() was, but to no avail.

**Comment by robertesler** [comment]
Hi @dimdieo.  It's definitely not a libpd issue.  I've had lots of problems using macOS's lock screen controls too.  Coming from Android it wasn't straight forward at all.  I wasn't entirely clear of your ultimate goal using the lock screen, but I was able to get the lock screen to control my app and sync with Bluetooth too.  
However, it was very difficult to sync the state of the lock screen UI.  The only way I could sync the stop/play button between the main UI, lock screen and Bluetooth controls was to directly call:
`self.audioController.active = YES;` or
 `self.audioController.active = NO;` using Obj-C for each command.  However, this would consistently crash the audio and I would have to restart the app.  So instead I gave up trying to sync between the different UI layers and figured anyone who uses it would just deal with it.  Instead I changed the text to display "PLAY" or "PAUSE", etc. and call a volume fade in Pd; [switch~] also worked.
I wasn't sure if this was the issue you are trying to solve, so if you did figure this out I am very curious.  I got the impression Apple assumes you are trying to just play media files (ideally from their media player API) if you want access to the lock screen (aka `MPNowPlayingInfoCenter`).   I don't know if anything has changed in iOS 17, this was for iOS 16.
